docs/architecture-decisions/2023-06-06_frontend-templates.md

```diff
@@ -0,0 +1,151 @@
+
+### Prerequisites
+
+I believe the following is a list of challenges that is common to all of the options and needs to be addressed:
+
+a) A set of new parameters for `.algokit.toml` file that will allow frontend codebases to quickly figure out the default path to generated typed clients. This will allow frontend templates to quickly import the generated typed clients and start interacting with the smart contract. Default templates provided by AlgoKit can rely on `copier` to pre-define the paths while an external frontend template can rely on the new parameters to figure out the path to the generated typed client.
+
+b) A new `dev` mode represented as a set of deployment configurations OR a new command for the algokit-cli that will allow a developer to Deploy Contract -> Spin up local frontend -> Interact with the contract via the frontend in sandbox mode. This will allow developers to quickly access, test and iterate on entire codebase e2e.
+
+### Option 1. Monorepo Approach
+
+In this approach, we can extend the existing templates to include frontend code. This would turn the template repositories into monorepos housing both frontend and backend code. This approach also implies tight coupling between the frontend and backend code, as they will live in the same repository. While sacrificing flexibility, I believe there is an equal amount of challenges in maintaining separate frontends and backends in separate repos. This approach is also consistent with the current approach of bundling smart contracts and typed clients in a single repository.
+
+#### Pros
+
+- The codebase is located in a single repository, improving code visibility and coordination.
+- Versioning is simpler as all code lives in one repo.
+- Ability to have a very polished frontend client tailored specifically for the smart contract template.
+- Collaboration among maintainers of official templates becomes easier as all code lives in one repo and it is easier to coordinate and observe changes.
+
+#### Cons
+
+- Without proper supervision over automation, the repository may become large and complex, leading to slower CI/CD pipeline execution times.
+- It might violate the principle of modularity and loosely coupled components, which can be beneficial for parallel development and reduced complexity.
+
+#### Examples
+
+Folder structure (ref official beaker template):
+
+```md
+.
+├── README.md
+├── copier.yaml
+├── poetry.lock
+├── poetry.toml
+├── pyproject.toml
+├── template_content
+├── contracts
+├── frontend
+├── .algokit.toml // with new params that aid with frontend templating
+└── ...
+├── tests
+└── tests_generated
+```
+
+A good reference example of bundling frontends and backens in a single repo that relies on jinja: https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master
+
+---
+
+### Option 2. Separate templates repositories
```

Comment:
A variation on this may be to allow `algokit init` on an existing repo so that a frontend template can be applied to a repo that already contains a backend. The .alogkit.toml could be leveraged with the variables described earlier to help coordinate this.

Additionally we could provide template definitions that include multiple template repo references, and the CLI then essentially runs `algokit init` on each template. This would allow for a more streamlined experience without increasing the verbosity of the commands

Reply:
Agreed, ill resolve the conversation as there is a third revision now that focuses on the proposals that the team currently agrees with the most and adds more implementation specific decision points. Feel free to check the new revision